File: backend/dotnet-services/src/GccBond.Portfolio/Services/PortfolioService.cs
using GccBond.Shared.Interfaces;
using GccBond.Portfolio.DTOs;
using GccBond.Portfolio.Interfaces;
using GccBond.Shared.Infrastructure;
using GccBond.Shared.Models;
using Serilog;

namespace GccBond.Portfolio.Services;

/// <summary>
/// Portfolio business logic (FSD §6.3).
/// No SQL — all data access goes through IPortfolioRepository.
/// </summary>
public class PortfolioService : IPortfolioService
{
    private readonly IPortfolioRepository _repo;
    private readonly IEventBus            _eventBus;

    public PortfolioService(IPortfolioRepository repo, IEventBus eventBus)
    {
        _repo     = repo;
        _eventBus = eventBus;
    }

    public async Task<PortfolioSummaryResponse> GetSummaryAsync(Guid userId)
    {
        var raw  = await _repo.GetHoldingDetailsAsync(userId);
        var list = raw.Select(h => new HoldingDetailResponse
        {
            BondId              = (Guid)h.bond_id,
            BondName            = (string)h.bond_name,
            Isin                = (string)h.isin,
            Quantity            = (decimal)h.quantity,
            AvgBuyPrice         = (decimal)h.avg_buy_price,
            CurrentPrice        = (decimal)h.current_price,
            CurrentValue        = (decimal)h.current_value,
            UnrealizedPnl       = (decimal)h.unrealized_pnl,
            UnrealizedPnlPct    = (decimal)h.unrealized_pnl_pct,
            TotalCouponReceived = (decimal)h.total_coupon_received,
            MaturityDate        = (DateTime)h.maturity_date,
            CouponFrequency     = (string)h.coupon_frequency,
            CouponRate          = (decimal)h.coupon_rate,
        }).ToList();

        var totalValue  = list.Sum(h => h.CurrentValue);
        var totalCost   = list.Sum(h => h.Quantity * h.AvgBuyPrice);
        var totalPnl    = list.Sum(h => h.UnrealizedPnl);
        var totalCoupons= list.Sum(h => h.TotalCouponReceived);
        var pnlPct      = totalCost > 0 ? Math.Round(totalPnl / totalCost * 100, 2) : 0;

        var withAlloc = list.Select(h => h with
        {
            Allocation = totalValue > 0 ? Math.Round(h.CurrentValue / totalValue * 100, 2) : 0,
        }).ToList();

        return new PortfolioSummaryResponse
        {
            TotalValue          = totalValue,
            TotalCost           = totalCost,
            TotalUnrealizedPnl  = totalPnl,
            TotalCouponReceived = totalCoupons,
            PnlPercentage       = pnlPct,
            Holdings            = withAlloc,
        };
    }

    public Task<IEnumerable<PortfolioHolding>> GetHoldingsAsync(Guid userId)
        => _repo.GetRawHoldingsAsync(userId);

    public Task<IEnumerable<dynamic>> GetCouponCalendarAsync(Guid userId)
        => _repo.GetCouponCalendarAsync(userId);

    public async Task ProcessCouponPaymentsAsync()
    {
        Log.Information("Processing coupon payments for {Date}", DateTime.UtcNow.Date);

        var dueBonds = await _repo.GetDueCouponsAsync(DateTime.UtcNow);

        foreach (var bond in dueBonds)
        {
            var holders = await _repo.GetHoldersByBondAsync((Guid)bond.bond_id);
            var m = GetPaymentsPerYear((string)bond.coupon_frequency);
            var couponPerUnit = Math.Round((decimal)bond.face_value * (decimal)bond.coupon_rate / m, 4);

            foreach (var holder in holders)
            {
                var amount = couponPerUnit * (decimal)holder.quantity;
                await _repo.ProcessCouponPaymentAsync(
                    (Guid)holder.user_id, (Guid)bond.bond_id,
                    amount, (string)bond.currency, (string)bond.name);

                await _eventBus.PublishAsync(Events.CouponPaid, new
                {
                    user_id  = holder.user_id.ToString(),
                    bond_id  = bond.bond_id.ToString(),
                    amount   = amount,
                    currency = (string)bond.currency,
                });
            }

            await _repo.MarkCouponPaidAsync((Guid)bond.bond_id, DateTime.UtcNow);
            Log.Information("Coupon processed for bond {BondId}: {Count} holders", bond.bond_id, holders.Count());
        }
    }

    private static int GetPaymentsPerYear(string frequency) => frequency switch
    {
        "Annual"     => 1,
        "SemiAnnual" => 2,
        "Quarterly"  => 4,
        _            => 1,
    };
}

File: backend/node-api/src/modules/portfolio/portfolio.service.ts
import { PortfolioRepository } from './portfolio.repository';
import { PortfolioHolding, PortfolioSummary, CouponEvent } from './portfolio.types';

export class PortfolioService {
  constructor(private readonly repo: PortfolioRepository) {}

  async getHoldings(userId: string): Promise<PortfolioHolding[]> {
    return this.repo.findHoldingsByUser(userId);
  }

  async getSummary(userId: string): Promise<PortfolioSummary> {
    const holdings = await this.repo.findHoldingsByUser(userId);

    const totalValue          = holdings.reduce((sum, h) => sum + h.currentValue,        0);
    const totalCost           = holdings.reduce((sum, h) => sum + h.avgBuyPrice * h.quantity, 0);
    const unrealizedPnl       = holdings.reduce((sum, h) => sum + h.unrealizedPnl,       0);
    const totalCouponReceived = holdings.reduce((sum, h) => sum + h.totalCouponReceived,  0);

    return {
      totalValue,
      totalCost,
      unrealizedPnl,
      totalCouponReceived,
      holdingsCount: holdings.length,
      currency:      'AED',
    };
  }

  async getCouponCalendar(userId: string): Promise<CouponEvent[]> {
    const rows = await this.repo.findCouponCalendarData(userId);
    const events: CouponEvent[] = [];
    const today   = new Date(); today.setHours(0, 0, 0, 0);
    const ceiling = new Date(today); ceiling.setFullYear(ceiling.getFullYear() + 2);

    const freqMap: Record<string, number> = {
      Annual: 1, SemiAnnual: 2, Quarterly: 4,
    };

    for (const row of rows) {
      const paymentsPerYear   = freqMap[row.couponFrequency] ?? 1;
      const intervalMonths    = Math.round(12 / paymentsPerYear);
      const maturity          = new Date(row.maturityDate);
      const couponPerPayment  = parseFloat(
        ((row.couponRate / 100 / paymentsPerYear) * row.quantity * row.faceValue).toFixed(2),
      );

      // Anchor on maturity date and walk back to find first upcoming coupon
      let d = new Date(maturity.getFullYear(), maturity.getMonth(), maturity.getDate());
      while (d > today) {
        d = new Date(d.getFullYear(), d.getMonth() - intervalMonths, d.getDate());
      }
      d = new Date(d.getFullYear(), d.getMonth() + intervalMonths, d.getDate());

      while (d <= maturity && d <= ceiling) {
        if (d >= today) {
          events.push({
            bondId:     row.bondId,
            bondName:   row.bondName,
            isin:       row.isin,
            date:       d.toISOString().split('T')[0],
            amount:     couponPerPayment,
            couponRate: row.couponRate,
          });
        }
        d = new Date(d.getFullYear(), d.getMonth() + intervalMonths, d.getDate());
      }
    }

    return events.sort((a, b) => a.date.localeCompare(b.date));
  }
}
